pkg/client/unversioned/request.go

```diff
@@ -164,6 +169,12 @@ func (r *Request) SubResource(subresources ...string) *Request {
 		r.err = fmt.Errorf("subresource already set to %q, cannot change to %q", r.resource, subresource)
 		return r
 	}
+	for _, s := range subresources {
```

Comment:
> Fixing this in the client is at best an optimization for non-adversarial clients, and I worry it will lead to us leaving a vulnerability in the server.

You need both client-side and server-side checks. Server-side URL normalization we get for free(ish) from ServeMux, and we obviously want to defer as much validation to the server as possible, but it only makes sense to prevent the client from building requests that are invalid, yet will appear valid to the server (like path segments that will collapse or traverse).

> I would at least like to not import the entire api/validation package

I can copy the check into this package if you like.

> I think it might be simpler to check the path once in the function that builds the final URL.

The checks need to remain on the individual segments. Once you combine segments into the final URL, you lose to ability to detect invalid resource names like "foo/bar" (not to mention path.Join(...) does a Clean, which also masks traversal)